segmentstore/server/src/main/java/io/pravega/segmentstore/server/CacheManager.java

```diff
@@ -47,17 +49,15 @@
 @ThreadSafe
 public class CacheManager extends AbstractScheduledService implements AutoCloseable {
     //region Members
-
+    private static final int CACHE_FULL_RETRY_BASE_MILLIS = 50;
     private static final String TRACE_OBJECT_ID = "CacheManager";
     @GuardedBy("lock")
     private final Collection<Client> clients;
     private final ScheduledExecutorService executorService;
-    @GuardedBy("lock")
-    private int currentGeneration;
-    @GuardedBy("lock")
-    private int oldestGeneration;
-    @GuardedBy("lock")
-    private CacheState lastCacheState;
+    private final AtomicInteger currentGeneration;
```

Comment:
As usual when changing from course locking to fine locking make double sure there aren't code paths that implicitly assume that another thread isn't going to change something in the mean time. For example here one might make the assumption that currentGeneration would not change in the middle of a method, or that current and oldest generation are different or that older is newer that current etc.

Reply:
Agreed. In this class, the only thing changing the internal state is `applyCachePolicyInternal`, which is guarded by the lock. I decided to make all these fields `AtomicXYZ` because I didn't want to acquire some lock just to query them. To be clear, `applyCachePolicyInternal` modifies these values, but they are queried in other places as well, which should be OK for thread safety purposes.

The only exception to this rule is `this.clients`, which keeps track of all the Cache Manager's Clients, and is a collection that is modified when new segments are activated or evicted from memory, hence out of the control of the Cache Manager. However, no external APIs are invoked while using the `clients` lock, so no chance for deadlock there.